src/python/pants/util/objects.py

```diff
@@ -585,3 +585,13 @@ def __repr__(self):
     return ('{type_constraint_type}({constraint!r})'
             .format(type_constraint_type=type(self).__name__,
                     constraint=self._constraint))
+
+
+hashable_collection_constraint = Exactly(tuple)
+
+
+class HashableTypedCollection(TypedCollection):
+  _iterable_constraint = hashable_collection_constraint
```

Comment:
What does setting `_iterable_constraint` do? If this does something on the super-class, it probably shouldn't start with an `_` because subclasses aren't meant to look at `_` fields.

Reply:
This does something on the superclass! I'll make it an `@classproperty` without an underscore and give it a docstring.

Reply:
Done!